lib/validator/validator.cpp

```diff
@@ -19,6 +20,69 @@ using namespace std::literals;
 namespace WasmEdge {
 namespace Validator {
 
+namespace details {
+static constexpr uint32_t MaxSubtypeDepth = 63;
+
+namespace {
+Expect<void>
+calculateSubtypeDepthRecursiveHelper(uint32_t Index, uint32_t Depth,
+                                     std::set<uint32_t> &VisitedSet,
+                                     FormChecker &Checker, uint32_t TypeIdx) {
+  if (VisitedSet.count(Index)) {
+    spdlog::error(ErrCode::Value::InvalidSubType);
+    spdlog::error("Cycle detected in subtype hierarchy for type {}.", Index);
+    return Unexpect(ErrCode::Value::InvalidSubType);
+  }
+
+  if (Depth >= MaxSubtypeDepth) {
+    spdlog::error(ErrCode::Value::InvalidSubType);
+    spdlog::error("subtype depth for Type section's {}th signature exceeded "
+                  "the limits of {}",
+                  TypeIdx + 1, MaxSubtypeDepth);
+    return Unexpect(ErrCode::Value::InvalidSubType);
+  }
+
+  VisitedSet.insert(Index);
+  const auto &TypeVec = Checker.getTypes();
+
+  if (unlikely(Index >= TypeVec.size())) {
+    spdlog::error(ErrCode::Value::InvalidSubType);
+    spdlog::error(
+        "    Type index {} is out of bounds for defined types (size: {}).",
+        Index, TypeVec.size());
+    return Unexpect(ErrCode::Value::InvalidSubType);
+  }
+
+  const auto &Type = *TypeVec[Index];
+
+  for (const auto SuperIdx : Type.getSuperTypeIndices()) {
+    if (unlikely(SuperIdx >= TypeVec.size())) {
```

Comment:
Not necessary to check the defined type bound.
Because all types and their super type indices added into the form checker are valid.